File: src/lib/wow/archive/casc/casc-source.ts
/**
 * Base CASC source, ported from wow.export (src/js/casc/casc-source.js).
 * UI couplings (Vue watches, listfile tab filters) are replaced by wowConfig
 * and the runtime load-func registry.
 */
import { write } from '@/lib/wow/log';
import { createProgress, Progress, runLoadFuncs } from '@/lib/wow/server/runtime';

import { BufferWrapper } from '../../formats/buffer';
import { wowConfig } from '../../server/config';
import { BLTEReader } from './blte-reader';
import { BuildCache } from './build-cache';
import {
  asCascKey, CascKey, cascKeyFromHex, cascKeyToHex,
} from './casc-key';
import { CDNConfigEntries } from './cdn-config';
import { ContentFlags } from './content-flags';
import { InstallManifest } from './install-manifest';
import * as listfile from './listfile';
import { LocaleFlags } from './locale-flags';
import { VersionConfigEntry } from './version-config';

const ENC_MAGIC = 0x4E45;
const ROOT_MAGIC = 0x4D465354;

export interface RootType {
  contentFlags: number;
  localeFlags: number;
}

/** Map of rootTypeIdx -> content key (compact binary). */
export type RootEntry = Map<number, CascKey>;

export abstract class CASC {
  encodingSizes = new Map<CascKey, number>();

  encodingKeys = new Map<CascKey, CascKey>();

  rootTypes: RootType[] = [];

  rootEntries = new Map<number, RootEntry>();

  isRemote: boolean;

  isLoaded = false;

  locale: number;

  progress: Progress = createProgress();

  buildConfig!: CDNConfigEntries;

  cdnConfig!: CDNConfigEntries;

  builds: VersionConfigEntry[] = [];

  build!: VersionConfigEntry;

  abstract cache: BuildCache;

  constructor(isRemote = false) {
    this.isRemote = isRemote;

    const locale = wowConfig.cascLocale;
    if (!Number.isNaN(locale)) {
      this.locale = locale;
    } else {
      write('Invalid locale set in configuration, defaulting to enUS');
      this.locale = LocaleFlags.enUS;
    }
  }

  abstract getFile(
    fileDataID: number,
    partialDecrypt?: boolean,
    suppressLog?: boolean,
    supportFallback?: boolean,
    forceFallback?: boolean,
    contentKey?: CascKey | null,
  ): Promise<BLTEReader>;

  abstract getDataFile(key: string): Promise<BufferWrapper>;

  abstract formatCDNKey(key: string): string;

  abstract getBuildName(): string;

  abstract getBuildKey(): string;

  abstract getProductList(): string[];

  abstract load(buildIndex: number): Promise<void>;

  /** Clear parsed index state before (re)loading a build. */
  resetForLoad(): void {
    this.encodingSizes.clear();
    this.encodingKeys.clear();
    this.rootTypes.length = 0;
    this.rootEntries.clear();
    this.isLoaded = false;
  }

  /** Provides an array of fileDataIDs that match the current locale. */
  getValidRootEntries(): number[] {
    const entries: number[] = [];

    for (const [fileDataID, entry] of this.rootEntries.entries()) {
      let include = false;

      for (const rootTypeIdx of entry.keys()) {
        const rootType = this.rootTypes[rootTypeIdx];
        if ((rootType.localeFlags & this.locale) && ((rootType.contentFlags & ContentFlags.LowViolence) === 0)) {
          include = true;
          break;
        }
      }

      if (include) entries.push(fileDataID);
    }

    return entries;
  }

  /** Retrieves the install manifest for this CASC instance. */
  async getInstallManifest(): Promise<InstallManifest> {
    const installKeys = this.buildConfig.install.split(' ');
    const installKey = installKeys.length === 1
      ? this.encodingKeys.get(cascKeyFromHex(installKeys[0]))!
      : cascKeyFromHex(installKeys[1]);

    const installKeyHex = cascKeyToHex(installKey);
    const raw = this.isRemote
      ? await this.getDataFile(this.formatCDNKey(installKeyHex))
      : await (this as unknown as { getDataFileWithRemoteFallback(key: string): Promise<BufferWrapper> }).getDataFileWithRemoteFallback(installKeyHex);
    const manifest = new BLTEReader(raw, installKeyHex);

    return new InstallManifest(manifest);
  }

  /** Check if a file exists by its fileDataID. */
  fileExists(fileDataID: number): boolean {
    const root = this.rootEntries.get(fileDataID);
    if (root === undefined) return false;

    for (const [rootTypeIdx] of root.entries()) {
      const rootType = this.rootTypes[rootTypeIdx];
      if ((rootType.localeFlags & this.locale) && ((rootType.contentFlags & ContentFlags.LowViolence) === 0)) return true;
    }

    return false;
  }

  /**
   * Obtain the encoding key for a file by its fileDataID.
   * (Underlying implementation of getFile; CASCLocal and CASCRemote implement readers.)
   */
  protected getEncodingKey(fileDataID: number): CascKey {
    const root = this.rootEntries.get(fileDataID);
    if (root === undefined) throw new Error(`fileDataID does not exist in root: ${fileDataID}`);

    let contentKey: CascKey | null = null;
    for (const [rootTypeIdx, key] of root.entries()) {
      const rootType = this.rootTypes[rootTypeIdx];

      // Select the first root entry that has a matching locale and no LowViolence flag set.
      if ((rootType.localeFlags & this.locale) && ((rootType.contentFlags & ContentFlags.LowViolence) === 0)) {
        contentKey = key;
        break;
      }
    }

    if (contentKey === null) throw new Error(`No root entry found for locale: ${this.locale}`);

    return this.getEncodingKeyForContentKey(contentKey);
  }

  getEncodingKeyForContentKey(contentKey: CascKey): CascKey {
    const key = asCascKey(contentKey);
    const encodingKey = this.encodingKeys.get(key);
    if (encodingKey === undefined) throw new Error(`No encoding entry found: ${cascKeyToHex(key)}`);

    return encodingKey;
  }

  /**
   * Obtain a file by a filename.
   * fileName must exist in the loaded listfile.
   */
  async getFileByName(
    fileName: string,
    partialDecrypt = false,
    suppressLog = false,
    supportFallback = true,
    forceFallback = false,
  ): Promise<BLTEReader> {
    let fileDataID: number | undefined;

    // If filename is "unknown/<fdid>", skip listfile lookup
    if (fileName.startsWith('unknown/') && !fileName.includes('.')) fileDataID = parseInt(fileName.split('/')[1], 10);
    else fileDataID = listfile.getByFilename(fileName);

    if (fileDataID === undefined) throw new Error(`File not mapping in listfile: ${fileName}`);

    return this.getFile(fileDataID, partialDecrypt, suppressLog, supportFallback, forceFallback);
  }

  /**
   * Prepare listfile data before loading.
   * Ensures preloading is complete to avoid race conditions.
   */
  async prepareListfile(): Promise<void> {
    await this.progress.step('Preparing listfiles...');
    await listfile.prepareListfile();
  }

  /** Load the listfile for selected build. */
  async loadListfile(_buildKey?: string): Promise<void> {
    await this.progress.step('Loading listfiles');
    listfile.applyPreload(this.rootEntries);
  }

  /**
   * Initialize external components as part of the CASC load process
   * (DB caches and other registered load functions).
   */
  async initializeComponents(): Promise<void> {
    await this.progress.step('Initializing components');
    await runLoadFuncs();
  }

  /** Parse entries from a root file. Returns the entry count. */
  parseRootFile(data: BufferWrapper, hash: string): number {
    const root = new BLTEReader(data, hash);

    const magic = root.readUInt32LE();
    const rootTypes = this.rootTypes;
    const rootEntries = this.rootEntries;

    if (magic === ROOT_MAGIC) { // 8.2
      let headerSize = root.readUInt32LE();
      let version = root.readUInt32LE();

      if (headerSize !== 0x18) {
        version = 0; // This will break with future header size increases.
      } else if (version !== 1 && version !== 2) {
        throw new Error(`Unknown root version: ${version}`);
      }

      let totalFileCount: number;
      let namedFileCount: number;

      if (version === 0) {
        totalFileCount = headerSize;
        namedFileCount = version;
        headerSize = 12;
      } else {
        totalFileCount = root.readUInt32LE();
        namedFileCount = root.readUInt32LE();
      }

      root.seek(headerSize);

      const allowNamelessFiles = totalFileCount !== namedFileCount;

      while (root.remainingBytes > 0) {
        const numRecords = root.readUInt32LE();

        let contentFlags = 0;
        let localeFlags = 0;

        if (version === 0 || version === 1) {
          contentFlags = root.readUInt32LE();
          localeFlags = root.readUInt32LE();
        } else if (version === 2) {
          localeFlags = root.readUInt32LE();
          const cflags1 = root.readUInt32LE();
          const cflags2 = root.readUInt32LE();
          const cflags3 = root.readUInt8();
          contentFlags = cflags1 | cflags2 | (cflags3 << 17);
        }

        const fileDataIDs = new Array<number>(numRecords);

        let fileDataID = 0;
        for (let i = 0; i < numRecords; i++) {
          const nextID = fileDataID + root.readInt32LE();
          fileDataIDs[i] = nextID;
          fileDataID = nextID + 1;
        }

        // Parse MD5 content keys for entries.
        for (let i = 0; i < numRecords; i++) {
          const fdid = fileDataIDs[i];
          let entry = rootEntries.get(fdid);

          if (!entry) {
            entry = new Map();
            rootEntries.set(fdid, entry);
          }

          entry.set(rootTypes.length, root.readBinaryKey(16));
        }

        // Skip lookup hashes for entries.
        if (!(allowNamelessFiles && (contentFlags & ContentFlags.NoNameHash))) root.move(8 * numRecords);

        // Push the rootType after parsing the block so that
        // rootTypes.length can be used for the type index above.
        rootTypes.push({ contentFlags, localeFlags });
      }
    } else { // Classic
      root.seek(0);
      while (root.remainingBytes > 0) {
        const numRecords = root.readUInt32LE();

        const contentFlags = root.readUInt32LE();
        const localeFlags = root.readUInt32LE();

        const fileDataIDs = new Array<number>(numRecords);

        let fileDataID = 0;
        for (let i = 0; i < numRecords; i++) {
          const nextID = fileDataID + root.readInt32LE();
          fileDataIDs[i] = nextID;
          fileDataID = nextID + 1;
        }

        // Parse MD5 content keys for entries.
        for (let i = 0; i < numRecords; i++) {
          const key = root.readBinaryKey(16);
          root.move(8); // hash

          const fdid = fileDataIDs[i];
          let entry = rootEntries.get(fdid);

          if (!entry) {
            entry = new Map();
            rootEntries.set(fdid, entry);
          }

          entry.set(rootTypes.length, key);
        }

        // Push the rootType after parsing the block so that
        // rootTypes.length can be used for the type index above.
        rootTypes.push({ contentFlags, localeFlags });
      }
    }

    return rootEntries.size;
  }

  /** Parse entries from an encoding file. */
  parseEncodingFile(data: BufferWrapper, hash: string): void {
    const encodingSizes = this.encodingSizes;
    const encodingKeys = this.encodingKeys;

    const encoding = new BLTEReader(data, hash);

    const magic = encoding.readUInt16LE();
    if (magic !== ENC_MAGIC) throw new Error(`Invalid encoding magic: ${magic}`);

    encoding.move(1); // version
    const hashSizeCKey = encoding.readUInt8();
    const hashSizeEKey = encoding.readUInt8();
    const cKeyPageSize = encoding.readInt16BE() * 1024;
    encoding.move(2); // eKeyPageSize
    const cKeyPageCount = encoding.readInt32BE();
    encoding.move(4 + 1); // eKeyPageCount + unk11
    const specBlockSize = encoding.readInt32BE();

    encoding.move(specBlockSize + (cKeyPageCount * (hashSizeCKey + 16)));

    const pagesStart = encoding.offset;
    for (let i = 0; i < cKeyPageCount; i++) {
      const pageStart = pagesStart + (cKeyPageSize * i);
      encoding.seek(pageStart);

      while (encoding.offset < (pageStart + pagesStart)) {
        const keysCount = encoding.readUInt8();
        if (keysCount === 0) break;

        const size = encoding.readInt40BE();
        const cKey = encoding.readBinaryKey(hashSizeCKey);

        encodingSizes.set(cKey, size);
        encodingKeys.set(cKey, encoding.readBinaryKey(hashSizeEKey));

        encoding.move(hashSizeEKey * (keysCount - 1));
      }
    }
  }

  /**
   * Run any necessary clean-up once a CASC instance is no longer needed.
   */
  cleanup(): void {
    // No config watches in the headless port.
  }
}

export default CASC;

File: golang/internal/wow/archive/casc/casc_source.go
package casc

import (
	"fmt"
	"strconv"
	"strings"

	"github.com/pqhuy98/wow-converter/internal/buffer"
	"github.com/pqhuy98/wow-converter/internal/wow/server"
	"github.com/pqhuy98/wow-converter/internal/wow/log"
	"github.com/pqhuy98/wow-converter/internal/wow/runtime"
)

const (
	encMagic  = 0x4E45
	rootMagic = 0x4D465354
)

// RootType describes a CASC root type entry.
type RootType struct {
	ContentFlags uint32
	LocaleFlags  uint32
}

// RootEntry maps rootTypeIdx -> content key.
type RootEntry map[int]CascKey

// CASC is the interface for local and remote CASC sources.
type CASC interface {
	GetFile(fileDataID int, partialDecrypt bool, suppressLog bool, supportFallback bool, forceFallback bool, contentKey CascKey) (*BLTEReader, error)
	GetDataFile(key string) (*buffer.Buffer, error)
	FormatCDNKey(key string) string
	GetBuildName() string
	GetBuildKey() string
	GetProductList() []string
	Load(buildIndex int) error
	Cache() *BuildCache
	IsRemote() bool
	IsLoaded() bool
	Locale() int
	EncodingEntryCount() int
	RootTypes() []RootType
	RootEntries() map[int]RootEntry
	BuildConfig() CDNConfigEntries
	CDNConfig() CDNConfigEntries
	Builds() []VersionConfigEntry
	Build() VersionConfigEntry
}

// BaseCASC contains shared CASC state and logic.
type BaseCASC struct {
	EncodingTable    *EncodingTable
	RootTypesList    []RootType
	RootEntriesMap   map[int]RootEntry
	IsRemoteSource   bool
	Loaded           bool
	LocaleValue      int
	ProgressTracker  runtime.Progress
	BuildConfigData  CDNConfigEntries
	CDNConfigData    CDNConfigEntries
	BuildsList       []VersionConfigEntry
	CurrentBuild     VersionConfigEntry
	CacheStore       *BuildCache
}

// NewBaseCASC creates base CASC state.
func NewBaseCASC(isRemote bool) *BaseCASC {
	locale := server.GetConfig().CascLocale
	if locale <= 0 {
		log.Write("Invalid locale set in configuration, defaulting to enUS")
		locale = LocaleEnUS
	}
	return &BaseCASC{
		EncodingTable:    newEncodingTable(16, 16),
		RootEntriesMap:   map[int]RootEntry{},
		IsRemoteSource:   isRemote,
		LocaleValue:      locale,
		ProgressTracker:  runtime.CreateProgress(0),
	}
}

func (b *BaseCASC) IsRemote() bool            { return b.IsRemoteSource }
func (b *BaseCASC) IsLoaded() bool            { return b.Loaded }
func (b *BaseCASC) Locale() int               { return b.LocaleValue }
func (b *BaseCASC) EncodingEntryCount() int { return b.EncodingTable.len() }
func (b *BaseCASC) RootTypes() []RootType     { return b.RootTypesList }
func (b *BaseCASC) RootEntries() map[int]RootEntry { return b.RootEntriesMap }
func (b *BaseCASC) BuildConfig() CDNConfigEntries { return b.BuildConfigData }
func (b *BaseCASC) CDNConfig() CDNConfigEntries { return b.CDNConfigData }
func (b *BaseCASC) Builds() []VersionConfigEntry { return b.BuildsList }
func (b *BaseCASC) Build() VersionConfigEntry { return b.CurrentBuild }
func (b *BaseCASC) Cache() *BuildCache        { return b.CacheStore }

// ResetForLoad clears parsed index state before reloading.
func (b *BaseCASC) ResetForLoad() {
	b.EncodingTable.reset()
	b.RootTypesList = nil
	b.RootEntriesMap = map[int]RootEntry{}
	b.Loaded = false
}

// GetValidRootEntries returns fileDataIDs matching the current locale.
func (b *BaseCASC) GetValidRootEntries() []int {
	var entries []int
	for fileDataID, entry := range b.RootEntriesMap {
		include := false
		for rootTypeIdx := range entry {
			rootType := b.RootTypesList[rootTypeIdx]
			if (rootType.LocaleFlags&uint32(b.LocaleValue)) != 0 && (rootType.ContentFlags&uint32(ContentLowViolence)) == 0 {
				include = true
				break
			}
		}
		if include {
			entries = append(entries, fileDataID)
		}
	}
	return entries
}

// GetInstallManifest retrieves the install manifest for this CASC instance.
func (b *BaseCASC) GetInstallManifest(getter func(key string) (*buffer.Buffer, error), formatCDNKey func(string) string) (*InstallManifest, error) {
	installKeys := strings.Fields(b.BuildConfigData["install"])
	var installKey CascKey
	if len(installKeys) == 1 {
		encKey, ok := b.EncodingTable.lookupEncodingKey(CascKeyFromHex(installKeys[0]))
		if !ok {
			return nil, fmt.Errorf("no encoding entry found for install key")
		}
		installKey = encKey
	} else {
		installKey = CascKeyFromHex(installKeys[1])
	}
	installKeyHex := CascKeyToHex(installKey)
	var raw *buffer.Buffer
	var err error
	if b.IsRemoteSource {
		raw, err = getter(formatCDNKey(installKeyHex))
	} else {
		raw, err = getter(installKeyHex)
	}
	if err != nil {
		return nil, err
	}
	reader, err := NewBLTEReader(raw, installKeyHex, false)
	if err != nil {
		return nil, err
	}
	return NewInstallManifest(reader)
}

// FileExists checks if a file exists by fileDataID.
func (b *BaseCASC) FileExists(fileDataID int) bool {
	root, ok := b.RootEntriesMap[fileDataID]
	if !ok {
		return false
	}
	return b.selectRootContentKey(root) != ""
}

func (b *BaseCASC) selectRootContentKey(root RootEntry) CascKey {
	pick := func(requireLocale, skipLowViolence bool) CascKey {
		for rootTypeIdx, key := range root {
			rootType := b.RootTypesList[rootTypeIdx]
			if skipLowViolence && (rootType.ContentFlags&uint32(ContentLowViolence)) != 0 {
				continue
			}
			if (rootType.ContentFlags & uint32(ContentDoNotLoad)) != 0 {
				continue
			}
			if requireLocale && (rootType.LocaleFlags&uint32(b.LocaleValue)) == 0 {
				continue
			}
			return key
		}
		return ""
	}
	if key := pick(true, true); key != "" {
		return key
	}
	if key := pick(true, false); key != "" {
		return key
	}
	if key := pick(false, true); key != "" {
		return key
	}
	return pick(false, false)
}

// GetEncodingKey obtains the encoding key for a fileDataID.
func (b *BaseCASC) GetEncodingKey(fileDataID int) (CascKey, error) {
	root, ok := b.RootEntriesMap[fileDataID]
	if !ok {
		return "", fmt.Errorf("fileDataID does not exist in root: %d", fileDataID)
	}
	contentKey := b.selectRootContentKey(root)
	if contentKey == "" {
		return "", fmt.Errorf("no root entry found for locale: %d", b.LocaleValue)
	}
	return b.GetEncodingKeyForContentKey(contentKey)
}

// GetEncodingKeyForContentKey maps a content key to an encoding key.
func (b *BaseCASC) GetEncodingKeyForContentKey(contentKey CascKey) (CascKey, error) {
	key := AsCascKey(string(contentKey))
	encodingKey, ok := b.EncodingTable.lookupEncodingKey(key)
	if !ok {
		return "", fmt.Errorf("no encoding entry found: %s", CascKeyToHex(key))
	}
	return encodingKey, nil
}

// GetFileByName obtains a file by filename.
func (b *BaseCASC) GetFileByName(source CASC, fileName string, partialDecrypt, suppressLog, supportFallback, forceFallback bool) (*BLTEReader, error) {
	var fileDataID int
	var ok bool
	if strings.HasPrefix(fileName, "unknown/") && !strings.Contains(fileName, ".") {
		parts := strings.Split(fileName, "/")
		if len(parts) > 1 {
			fileDataID, _ = strconv.Atoi(parts[1])
			ok = fileDataID != 0
		}
	} else {
		fileDataID, ok = GetByFilename(fileName)
	}
	if !ok {
		return nil, fmt.Errorf("file not mapping in listfile: %s", fileName)
	}
	return source.GetFile(fileDataID, partialDecrypt, suppressLog, supportFallback, forceFallback, "")
}

// PrepareListfile ensures listfile preloading is complete.
func (b *BaseCASC) PrepareListfile() error {
	if err := b.ProgressTracker.Step("Preparing listfiles..."); err != nil {
		return err
	}
	_, err := PrepareListfile()
	return err
}

// LoadListfile applies the preloaded listfile for the selected build.
func (b *BaseCASC) LoadListfile() error {
	if err := b.ProgressTracker.Step("Loading listfiles"); err != nil {
		return err
	}
	ApplyPreload(b.RootEntriesMap)
	return nil
}

// InitializeComponents runs registered load functions.
func (b *BaseCASC) InitializeComponents() error {
	if err := b.ProgressTracker.Step("Initializing components"); err != nil {
		return err
	}
	return runtime.RunLoadFuncs()
}

// ParseRootFile parses entries from a root file.
func (b *BaseCASC) ParseRootFile(data *buffer.Buffer, hash string) (int, error) {
	root, err := NewBLTEReader(data, hash, false)
	if err != nil {
		return 0, err
	}
	magic := root.ReadUInt32LE().(int64)
	if uint32(magic) == rootMagic {
		headerSize := int(root.ReadUInt32LE().(int64))
		version := int(root.ReadUInt32LE().(int64))
		if headerSize != 0x18 {
			version = 0
		} else if version != 1 && version != 2 {
			return 0, fmt.Errorf("unknown root version: %d", version)
		}
		var totalFileCount, namedFileCount int
		if version == 0 {
			totalFileCount = headerSize
			namedFileCount = version
			headerSize = 12
		} else {
			totalFileCount = int(root.ReadUInt32LE().(int64))
			namedFileCount = int(root.ReadUInt32LE().(int64))
		}
		root.Seek(headerSize)
		allowNamelessFiles := totalFileCount != namedFileCount
		for root.RemainingBytes() > 0 {
			numRecords := int(root.ReadUInt32LE().(int64))
			var contentFlags, localeFlags uint32
			if version == 0 || version == 1 {
				contentFlags = uint32(root.ReadUInt32LE().(int64))
				localeFlags = uint32(root.ReadUInt32LE().(int64))
			} else if version == 2 {
				localeFlags = uint32(root.ReadUInt32LE().(int64))
				cflags1 := uint32(root.ReadUInt32LE().(int64))
				cflags2 := uint32(root.ReadUInt32LE().(int64))
				cflags3 := uint32(root.ReadUInt8().(int64))
				contentFlags = cflags1 | cflags2 | (cflags3 << 17)
			}
			fileDataIDs := make([]int, numRecords)
			fileDataID := 0
			for i := 0; i < numRecords; i++ {
				nextID := fileDataID + int(root.ReadInt32LE().(int64))
				fileDataIDs[i] = nextID
				fileDataID = nextID + 1
			}
			for i := 0; i < numRecords; i++ {
				fdid := fileDataIDs[i]
				entry := b.RootEntriesMap[fdid]
				if entry == nil {
					entry = RootEntry{}
					b.RootEntriesMap[fdid] = entry
				}
				entry[len(b.RootTypesList)] = CascKey(root.ReadBinaryKey(16))
			}
			if !(allowNamelessFiles && (contentFlags&uint32(ContentNoNameHash)) != 0) {
				root.Move(8 * numRecords)
			}
			b.RootTypesList = append(b.RootTypesList, RootType{ContentFlags: contentFlags, LocaleFlags: localeFlags})
		}
	} else {
		root.Seek(0)
		for root.RemainingBytes() > 0 {
			numRecords := int(root.ReadUInt32LE().(int64))
			contentFlags := uint32(root.ReadUInt32LE().(int64))
			localeFlags := uint32(root.ReadUInt32LE().(int64))
			fileDataIDs := make([]int, numRecords)
			fileDataID := 0
			for i := 0; i < numRecords; i++ {
				nextID := fileDataID + int(root.ReadInt32LE().(int64))
				fileDataIDs[i] = nextID
				fileDataID = nextID + 1
			}
			for i := 0; i < numRecords; i++ {
				key := CascKey(root.ReadBinaryKey(16))
				root.Move(8)
				fdid := fileDataIDs[i]
				entry := b.RootEntriesMap[fdid]
				if entry == nil {
					entry = RootEntry{}
					b.RootEntriesMap[fdid] = entry
				}
				entry[len(b.RootTypesList)] = key
			}
			b.RootTypesList = append(b.RootTypesList, RootType{ContentFlags: contentFlags, LocaleFlags: localeFlags})
		}
	}
	return len(b.RootEntriesMap), nil
}

// ParseEncodingFile parses entries from an encoding file.
func (b *BaseCASC) ParseEncodingFile(data *buffer.Buffer, hash string) error {
	encoding, err := NewBLTEReader(data, hash, false)
	if err != nil {
		return err
	}
	magic := encoding.ReadUInt16LE().(int64)
	if uint16(magic) != encMagic {
		return fmt.Errorf("invalid encoding magic: %d", magic)
	}
	encoding.Move(1)
	hashSizeCKey := int(encoding.ReadUInt8().(int64))
	hashSizeEKey := int(encoding.ReadUInt8().(int64))
	b.EncodingTable.init(hashSizeCKey, hashSizeEKey)
	cKeyPageSize := int(encoding.ReadInt16BE().(int64)) * 1024
	encoding.Move(2)
	cKeyPageCount := int(encoding.ReadInt32BE().(int64))
	encoding.Move(4 + 1)
	specBlockSize := int(encoding.ReadInt32BE().(int64))
	encoding.Move(specBlockSize + (cKeyPageCount * (hashSizeCKey + 16)))
	pagesStart := encoding.Offset()
	for i := 0; i < cKeyPageCount; i++ {
		pageStart := pagesStart + (cKeyPageSize * i)
		encoding.Seek(pageStart)
		for encoding.Offset() < (pageStart + pagesStart) {
			keysCount := int(encoding.ReadUInt8().(int64))
			if keysCount == 0 {
				break
			}
			size := int(encoding.ReadInt40BE().(int64))
			cKey := CascKey(encoding.ReadBinaryKey(hashSizeCKey))
			eKey := CascKey(encoding.ReadBinaryKey(hashSizeEKey))
			b.EncodingTable.set(cKey, eKey, size)
			encoding.Move(hashSizeEKey * (keysCount - 1))
		}
	}
	return nil
}

// Cleanup runs cleanup once a CASC instance is no longer needed.
func (b *BaseCASC) Cleanup() {}

// ReadInt40BE reads a 40-bit big-endian integer from BLTEReader.
func (r *BLTEReader) ReadInt40BE(count ...int) any {
	n := 5
	if len(count) > 0 {
		n = 5 * count[0]
	}
	r.CheckBounds(n)
	return r.Buffer.ReadInt40BE(count...)
}

// ReadUInt8Array is a helper alias for multi-byte reads used by install manifest.
func ReadUInt8Array(b *buffer.Buffer, count int) []byte {
	vals := b.ReadUInt8(count).([]int64)
	out := make([]byte, len(vals))
	for i, v := range vals {
		out[i] = byte(v)
	}
	return out
}
